components/constellation/constellation.rs

```diff
@@ -1376,6 +1384,94 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
+            .. RequestInit::default()
+        };
+
+        let res_init = ResponseInit {
+            url: load_data.url.clone(),
+            .. ResponseInit::default()
+        };
+
+        // keep looping until we get a response that doesn't redirect
+        loop {
+            let (sender, receiver) = ipc::channel().expect("Failed to create IPC channel!");
+            if let Err(e) = self.public_resource_threads.send(
+                                CoreResourceMsg::FetchRedirect(req_init.clone(),
+                                res_init.clone(),
+                                sender)) {
+                warn!("Exit resource thread failed ({})", e);
+            }
+            match receiver.recv() {
+                Ok(recvd) => {
+                    let data = match recvd {
+                        FetchResponseMsg::ProcessResponse(res_metadata) => {
+                            match res_metadata {
+                                Ok(metadata) => {
+                                    match metadata {
+                                        FetchMetadata::Filtered {
+                                            filtered,
+                                            unsafe_: unsafe_metadata
+                                        } => Some((filtered, unsafe_metadata)),
+                                        _ => None
+                                    }
+                                },
+                                Err(_) => None,
+                            }
+                        },
+                        _ => break
+                    };
+                    match data {
+                        Some((FilteredMetadata::OpaqueRedirect, _)) => {
+                            if let Some((_, unsafe_metadata)) = data {
+                                match unsafe_metadata.headers {
+                                    Some(headers) => match headers.get::<Location>() {
+                                        Some(url) => {
+                                            match ServoUrl::parse(url) {
+                                                Ok(parsed_url) => req_init.url = parsed_url,
+                                                Err(e) => {
+                                                    warn!("Parsing URL {} failed ({}).", url, e);
+                                                    break
+                                                },
+                                            };
+                                        },
+                                        None => break,
+                                    },
+                                    None => break
+                                };
+                            };
+                        },
+                        _ => break
+                    };
+                },
+                Err(_) => break,
+            };
+        }
+
+        let msg = ConstellationControlMsg::FinalRequest(id, req_init);
```

Comment:
I'm not looking at the rest of the code in this method, because I believe that restructuring it to use an async fetch listener will change it significantly.